### PR TITLE
feat: od least squares solver

### PR DESCRIPTION
**Pull request by vishwa2710**
## Summary by CodeRabbit

- **New Features**
  - Introduced enhanced state and orbit estimation capabilities via updated Python interfaces, offering users improved astrodynamics computations.
- **Bug Fixes**
  - Refined error handling in estimation operations to better address edge cases and invalid inputs.
- **Tests**
  - Expanded automated test coverage to ensure the robustness and reliability of the new estimation functionality, including new tests for error handling and validation of outputs.

### CODE REVIEW

File: bindings/python/src/OpenSpaceToolkitAstrodynamicsPy/Estimator.cpp [hunk 1]
```diff
@@ -0,0 +1,14 @@
+/// Apache License 2.0
+
+#include <OpenSpaceToolkitAstrodynamicsPy/Estimator/ODLeastSquaresSolver.cpp>
+
+inline void OpenSpaceAstrodynamicsPy_Estimator(pybind11::module& aModule)
+{
+    using namespace pybind11;
+
+    // Create "data" python submodule
```
Comment (alex-liang3) [inline bindings/python/src/OpenSpaceToolkitAstrodynamicsPy/Estimator.cpp:9]:
```suggestion
    // Create "estimator" python submodule
```

File: bindings/python/src/OpenSpaceToolkitAstrodynamicsPy/Estimator.cpp [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    auto estimator = aModule.def_submodule("estimator");
+
+    // Add objects to python "data" submodules
```
Comment (alex-liang3) [inline bindings/python/src/OpenSpaceToolkitAstrodynamicsPy/Estimator.cpp:12]:
```suggestion
    // Add objects to python "estimator" submodules
```

**Review by alex-liang3 — APPROVED** [review]
👏 👏 👏

File: bindings/python/test/estimator/test_od_least_squares_solver.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+def initial_guess_state(
+    reference_states: list[State],
+) -> State:
+    return reference_states[0]
+
+
+@pytest.fixture
+def reference_states() -> list[State]:
+    return generate_states_from_dataframe(
+        pd.read_csv("/app/test/OpenSpaceToolkit/Astrodynamics/Estimator/gnss_data.csv"),
```
Comment (apaletta3) [inline bindings/python/test/estimator/test_od_least_squares_solver.py:70]: hmmm should we use the gps data for the python tests? I think we should just keep the actual unit tests in cpp and not in the bindings no?

File: include/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.hpp [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+{
+   public:
+    class Analysis
+    {
+       public:
+        /// @brief Constructor
+        Analysis(const State& aDeterminedState, const LeastSquaresSolver::Analysis& anAnalysis);
+
+        /// @brief Destructor
+        ~Analysis() = default;
```
Comment (apaletta3) [inline include/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.hpp:55]: I don' think we do this very often in OSTk, not sure its needed

File: include/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.hpp [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    /// @param anEstimationCoordinateSubsets Coordinate subsets to estimate
+    Analysis estimateState(
+        const State& anInitialGuessState,
+        const Array<State>& aReferenceStateArray,
+        const Array<Shared<const CoordinateSubset>>& anEstimationCoordinateSubsets =
+            Array<Shared<const CoordinateSubset>>::Empty(),
+        const std::unordered_map<CoordinateSubset, VectorXd>& anInitialGuessSigmas = DEFAULT_INITIAL_GUESS_SIGMAS,
+        const std::unordered_map<CoordinateSubset, VectorXd>& aReferenceStateSigmas = DEFAULT_REFERENCE_STATE_SIGMAS
+    ) const;
+
```
Comment (apaletta3) [inline include/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.hpp:111]: Missing compute state and orbit residuals method

File: src/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.cpp [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+    // Estimation state builder: Contains the subsets to be estimated.
+    const StateBuilder estimationStateBuilder(
+        initialGuessState.accessFrame(),
+        anEstimationCoordinateSubsets.isEmpty() ? initialGuessState.getCoordinateSubsets()
+                                                : anEstimationCoordinateSubsets
+    );
+
+    // Validate estimator subsets
+    for (const auto& subset : estimationStateBuilder.getCoordinateSubsets())
```
Comment (apaletta3) [inline src/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.cpp:121]: Can also check for consistent frame here

File: include/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.hpp [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    static const Shared<const Frame> EstimationFrameSPtr;
+
+    /// @brief Constructor
+    ///
+    /// @param anEnvironment Environment
+    /// @param aNumericalSolver Numerical solver
+    /// @param aSolver Least squares solver
+    ODLeastSquaresSolver(
+        const Environment& anEnvironment,
```
Comment (phc1990) [inline include/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.hpp:82]: Should anEnvironment also provide a default?

File: src/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.cpp [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    const State& anInitialGuessState,
+    const Array<State>& aReferenceStateArray,
+    const Array<Shared<const CoordinateSubset>>& anEstimationCoordinateSubsets,
+    const std::unordered_map<CoordinateSubset, VectorXd>& anInitialGuessSigmas,
+    const std::unordered_map<CoordinateSubset, VectorXd>& aReferenceStateSigmas
+) const
+{
+    // Convert everything to an inertial frame for estimation
+
+    const State initialGuessState = anInitialGuessState.inFrame(EstimationFrameSPtr);
```
Comment (phc1990) [inline src/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.cpp:98]: For clarity

```suggestion
    const State initialGuessStateInFrame = anInitialGuessState.inFrame(EstimationFrameSPtr);
```

File: src/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.cpp [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+        estimationStateBuilder.reduce(initialGuessState),
+        aReferenceStateArray,
+        generateStatesCallback,
+        anInitialGuessSigmas,
+        aReferenceStateSigmas
+    );
+
+    // Expand solution state to full state
+    const State determinedState = propagationStateBuilder.expand(analysis.accessSolutionState(), initialGuessState)
+                                      .inFrame(anInitialGuessState.accessFrame());
```
Comment (phc1990) [inline src/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.cpp:149]: 1. should the Estimation Frame be something that can be set at construction? (and defaulted to GCRF otherwise)
2. agree with returning the determined state in the frame of the initial guess, my only comment is that the returned state and that of the analysis might be in different frames which could be counterintuitive

File: src/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.cpp [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+    const Array<Shared<const CoordinateSubset>>& anEstimationCoordinateSubsets,
+    const std::unordered_map<CoordinateSubset, VectorXd>& anInitialGuessSigmas,
+    const std::unordered_map<CoordinateSubset, VectorXd>& aReferenceStateSigmas
+) const
+{
+    // Convert everything to an inertial frame for estimation
+
+    const State initialGuessState = anInitialGuessState.inFrame(EstimationFrameSPtr);
+
+    const Array<State> referenceStates = aReferenceStateArray.map<State>(
```
Comment (phc1990) [inline src/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.cpp:100]: Same here, I think this would make it more readable:

```suggestion
    const Array<State> referenceStatesInFrame = aReferenceStateArray.map<State>(
```

File: src/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.cpp (hunk 6 above)
Comment (phc1990) [reply src/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.cpp:121]: what do you mean?

File: test/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.test.cpp [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+    {
+        // Test with sigmas
+        const auto analysis = odSolver_.estimateState(
+            referenceStates_[0], referenceStates_, {}, initialStateSigmas_, referenceStateSigmas_
+        );
+
+        EXPECT_EQ(analysis.getSolverAnalysis().getTerminationCriteria(), "RMS Update Threshold");
+        EXPECT_LT(analysis.getSolverAnalysis().getRmsError(), 50.0);
+    }
+}
```
Comment (phc1990) [inline test/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.test.cpp:165]: Could we add something like this:

```suggestion
    {
        // Test basic state estimator with miscellaneous frames
        const State initialGuessStateInTEME = referenceStates_[0].inFrame(Frame::TEME());
        const Array<State> referenceStatesInITRF = referenceStates_.map<State>(
            {
                return aState.inFrame(Frame::ITRF());
            }
        );
        
        const auto analysis = odSolver_.estimateState(referenceStates_[0], referenceStates_);
        const auto analysisInMiscellaneousFrames = odSolver_.estimateState(initialGuessStateInTEME, referenceStatesInITRF);

        EXPECT_EQ(analysisInMiscellaneousFrames.getSolverAnalysis().getTerminationCriteria(), analysis.getSolverAnalysis().getTerminationCriteria());
        EXPECT_EQ(analysisInMiscellaneousFrames.getSolverAnalysis().getRmsError(), analysis.getSolverAnalysis().getRmsError());  // Or assert that is close somehow
    }
}
```

**Review by phc1990 — APPROVED** [review]
Nice stuff 👍

File: src/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.cpp (hunk 6 above)
Comment (apaletta3) [reply src/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.cpp:121]: sorry nvm

File: include/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.hpp (hunk 7 above)
Comment (vishwa2710) [reply include/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.hpp:82]: Hmmmmmmmm I think yeah we could do that. We should update the Environment in Physics to not be a full sized EGM2008 model though

File: include/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.hpp (hunk 5 above)
Comment (vishwa2710) [reply include/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.hpp:111]: As discussed these will be added later due to the complexity of the signatures

File: bindings/python/test/estimator/test_od_least_squares_solver.py (hunk 3 above)
Comment (vishwa2710) [reply bindings/python/test/estimator/test_od_least_squares_solver.py:70]: It was easier than creating fake states lol

File: include/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.hpp [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+    ///
+    /// @param anEnvironment Environment, Defaults to Environment::Default()
+    /// @param aNumericalSolver Numerical solver, Defaults to NumericalSolver::Default()
+    /// @param aSolver Least squares solver, Defaults to LeastSquaresSolver::Default()
+    ODLeastSquaresSolver(
+        const Environment& anEnvironment = DEFAULT_ENVIRONMENT,
+        const NumericalSolver& aNumericalSolver = DEFAULT_NUMERICAL_SOLVER,
+        const LeastSquaresSolver& aSolver = DEFAULT_LEAST_SQUARES_SOLVER,
+        const Shared<const Frame>& anEstimationFrameSPtr = DEFAULT_ESTIMATION_FRAME
+    );
```
Comment (apaletta3) [inline include/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.hpp:82]: Do we really want this many defaults? constructing an ODLeastSquaresSolver() just like that feels a little weird...

**Review by apaletta3 — COMMENTED** [review]
Initial interface review

File: include/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.hpp [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+    ///
+    /// @param anInitialGuessState Initial guess state
+    /// @param anObservationArray Observations to fit against
+    /// @param anEstimationCoordinateSubsets Coordinate subsets to estimate. Defaults to empty array, in which case all
+    /// the coordinate subsets from the initial guess state are estimated
+    /// @param anInitialGuessSigmas Initial guess sigmas
+    /// @param anObservationSigmas Observation sigmas
+    ///
+    /// @return Analysis
+    Analysis estimateState(
```
Comment (apaletta3) [inline include/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.hpp:114]: `estimate`

File: include/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.hpp [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+        Analysis(const State& aDeterminedState, const LeastSquaresSolver::Analysis& anAnalysis);
+
+        /// @brief Print analysis
+        friend std::ostream& operator<<(std::ostream& anOutputStream, const Analysis& anAnalysis);
+
+        /// @brief Print analysis
+        void print(std::ostream& anOutputStream) const;
+
+        State determinedState;
+        LeastSquaresSolver::Analysis solverAnalysis;
```
[the remark this answers is not in the document]
Comment (vishwa2710) [reply include/OpenSpaceToolkit/Astrodynamics/Estimator/ODLeastSquaresSolver.hpp:69]: Since the estimatedState in the OD solver is the expanded version + in the same frame as the initial state, I think it's worth leaving this as is.
